mlflow/metrics/metric_definitions.py

```diff
@@ -373,6 +373,100 @@ def _fn(predictions, targets):
     return _fn
 
 
+def _expand_duplicate_retrieved_docs(predictions, targets):
+    counter = {}
+    expanded_predictions = []
+    expanded_targets = targets.copy()
+    for doc_id in predictions:
+        if doc_id not in counter:
+            counter[doc_id] = 1
+            expanded_predictions.append(doc_id)
+        else:
+            counter[doc_id] += 1
+            new_doc_id = (
+                f"{doc_id}_bc574ae_{counter[doc_id]}"  # adding a random string to avoid collisions
+            )
+            expanded_predictions.append(new_doc_id)
+            expanded_targets.add(new_doc_id)
+    return expanded_predictions, expanded_targets
+
+
+def _prepare_row_for_ndcg(predictions, targets):
+    """Prepare data one row from predictions and targets to y_score, y_true for ndcg calculation.
+
+    Args:
+        predictions: A list of strings of at most k doc IDs retrieved.
+        targets: A list of strings of ground-truth doc IDs.
+
+    Returns:
+        y_true : ndarray of shape (1, n_docs) Representing the ground-truth relevant docs.
+            n_docs is the number of unique docs in union of predictions and targets.
+        y_score : ndarray of shape (1, n_docs) Representing the retrieved docs.
+            n_docs is the number of unique docs in union of predictions and targets.
+    """
+    # sklearn does an internal sort of y_score, so to preserve the order of our retrieved
+    # docs, we need to modify the relevance value slightly
+    eps = 1e-6
+    targets = set(targets)
```

Comment:
nit: only need either this line or [this line](https://github.com/mlflow/mlflow/pull/10284/files#diff-4d216618e4891c4fb5751f7698964f0ea2fcffb184fd915877e90be401f538c6R379) because both will create a new set without mutating the underlying array

Reply:
Oh, nice catch, I remember we discussed this.